### PR TITLE
feat: add deploy script

### PR DESCRIPTION
**Pull request by ftupas**
Time spent on this PR: 0.5

Resolves: #112 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

This pr does the following:
- add deployment script using [starknet-rs](https://github.com/xJonathanLEI/starknet-rs)
- update and format README

Example contract deployment

https://testnet.starkscan.co/contract/0x0324c28298956a3084160ff02177c5da7a6bf0228c862cd1749d2d75b30e05c6

image

# Does this introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: scripts/deploy.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+use starknet::contract::ContractFactory;
+use starknet::core::types::contract::SierraClass;
+use starknet::core::types::{BlockId, BlockTag, FieldElement, StarknetError};
+use starknet::providers::jsonrpc::{HttpTransport, JsonRpcClient};
+use starknet::providers::{MaybeUnknownErrorCode, Provider, ProviderError, StarknetErrorWithMessage};
+use starknet::signers::{LocalWallet, SigningKey};
+
+const BUILD_PATH_PREFIX: &str = "target/dev/yas_";
+// TODO: Update to New once account contracts are migrated to v1
+const ENCODING: ExecutionEncoding = ExecutionEncoding::Legacy;
```
Comment (ftupas) [inline scripts/deploy.rs:16]: Add todo for encoding

File: scripts/deploy.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    contract_name: &str,
+) -> Result<FieldElement> {
+    // Load the contract artifact.
+    let contract_artifact = contract_artifact(contract_name)?;
+
+    // Compute the contract class hash.
+    let class_hash = contract_artifact.class_hash()?;
+
+    // Declare the contract class if it is not already declared.
+    if !is_already_declared(account.provider(), &class_hash).await? {
```
Comment (ftupas) [inline scripts/deploy.rs:110]: call `declare` only if the class hash hasn't been declared yet

File: scripts/deploy.rs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    let account_address = FieldElement::from_hex_be(&env::var("ACCOUNT_ADDRESS").expect("ACCOUNT_ADDRESS not set"))
+        .expect("Invalid Account Address");
+    let account = initialize_starknet_account(signer, account_address).await?;
+
+    // Declare the contract classes if they are not already declared.
+    let pool_class_hash = declare_contract(&account, "YASPool").await?;
+    let factory_class_hash = declare_contract(&account, "YASFactory").await?;
+
+    // Instantiate the contract factory.
+    let salt = account.get_nonce().await?;
```
Comment (ftupas) [inline scripts/deploy.rs:138]: Use account `nonce` for salt

File: scripts/deploy.rs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    let contract_factory = ContractFactory::new(factory_class_hash, account);
+    let owner_address = FieldElement::from_hex_be(&env::var("OWNER_ADDRESS").expect("OWNER_ADDRESS not set"))
+        .expect("Invalid Owner Address");
+    let unique = true;
+    println!(
+        "==> Deploying Factory Contract\nOWNER_ADDRESS: {:#064x}\nPOOL_CLASS_HASH: {:#064x}\nSALT: {}\nUNIQUE: {}",
+        owner_address, pool_class_hash, salt, unique
+    );
+
+    let contract_deployment = contract_factory.deploy(vec![owner_address, pool_class_hash], salt, unique);
```
Comment (ftupas) [inline scripts/deploy.rs:148]: constructor arguments: `[owner, pool_class_hash]`

**Review by rcatalan98 — APPROVED** [review]
Hey! Incredible work thanks! Just small comment please resolve the merge conflicts in the README.

**Comment by ftupas** [comment]
> Hey! Incredible work thanks! Just small comment please resolve the merge conflicts in the README.

Thank you ser, pulled from main and resolved merge conflicts 🫡

**Review by dubzn — APPROVED** [review]
Great job ftupas :fire:! I wonder, perhaps more generally, if we only want the local deployment in the readme. I think it adds value to also include the deployment to the Starknet testnet since there is a section ([Setting up testnet smart wallet](https://github.com/lambdaclass/yet-another-swap#setting-up-a-testnet-smart-wallet)) for configuring/creating the wallet via starkli. 

What do you think @rcatalan98 ?

**Comment by rcatalan98** [comment]
> Great job ftupas 🔥! I wonder, perhaps more generally, if we only want the local deployment in the readme. I think it adds value to also include the deployment to the Starknet testnet since there is a section ([Setting up testnet smart wallet](https://github.com/lambdaclass/yet-another-swap#setting-up-a-testnet-smart-wallet)) for configuring/creating the wallet via starkli.
> 
> What do you think @rcatalan98 ?

For now it's ok to only have this for local testing, we would need a different guide for testnet deployment. So people can understand what are the "official" contracts.

**Pull request merged**